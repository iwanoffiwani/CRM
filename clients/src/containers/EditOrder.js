import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import axios from "axios";
import { fetchUpdateOrderList } from "../redux/actions";
import { makeStyles } from "@material-ui/core/styles";
import Box from "@material-ui/core/Box";
import Container from "@material-ui/core/Container";
import Typography from "@material-ui/core/Typography";
import TextField from "@material-ui/core/TextField";
import Button from "@material-ui/core/Button";
import Divider from "@material-ui/core/Divider";
import MenuItem from "@material-ui/core/MenuItem";
import Select from "@material-ui/core/Select";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import ListItemText from "@material-ui/core/ListItemText";
import green from "@material-ui/core/colors/green";

const useStyles = makeStyles(theme => ({
  wrapper: {
    display: "flex",
    padding: "2rem 0"
  },
  error: {
    margin: 0,
    marginBottom: theme.spacing(1)
  },
  success: {
    margin: 0,
    marginBottom: theme.spacing(1),
    color: green[500]
  },
  formControl: {
    margin: 0
  },
  textField: {
    margin: 0,
    marginBottom: theme.spacing(1)
  },
  title: {
    marginBottom: theme.spacing(4),
    fontWeight: "bold",
    fontSize: "1.5em"
  },
  divider: {
    marginTop: theme.spacing(4),
    marginBottom: theme.spacing(6)
  },
  selectEmpty: {
    marginTop: theme.spacing(2)
  },
  comment: {
    width: 720
  },
  list: {
    padding: 0,
    height: 500,
    overflowX: "hidden"
  },
  listItem: {
    paddingRight: 0,
    paddingLeft: 0
  },
  data: {
    display: "block",
    marginTop: theme.spacing(1)
  },
  hiddenInput: {
    visibility: "hidden"
  }
}));

export const EditOrder = props => {
  const classes = useStyles();

  // Поле changes содержит объект с полем prevState в которое падаем объект с данными заявки,
  // для последующего сохранения и отражения изменений.
  // Поле nextState остаеться пустым так как в него будут записываеться изменения
  const initialState = {
    response: {
      success: {},
      error: {}
    },
    edit: {
      ...props.editOrder
    },
    comment: "",
    changes: {
      prevState: {
        ...props.editOrder
      },
      nextState: {}
    }
  };

  const [state, setState] = useState(initialState);

  const changeNameHandler = e => {
    return setState({
      ...state,
      edit: {
        ...state.edit,
        name: e.target.value
      },
      changes: {
        ...state.changes,
        nextState: {
          ...state.changes.nextState,
          name: e.target.value
        }
      }
    });
  };

  const changeStatusHandler = e => {
    const { id: _id } = e._targetInst.stateNode;

    return setState({
      ...state,
      edit: {
        ...state.edit,
        status: {
          _id,
          name: e.target.value
        }
      },
      changes: {
        ...state.changes,
        nextState: {
          ...state.changes.nextState,
          status: e.target.value
        }
      }
    });
  };

  const changeFieldsHandler = e => {
    return setState({
      ...state,
      edit: {
        ...state.edit,
        fields: state.edit.fields.map(field => {
          if (field.name === e.target.name)
            return {
              ...field,
              value: e.target.value
            };
          else return field;
        })
      },
      changes: {
        ...state.changes,
        nextState: {
          ...state.changes.nextState,
          fields: state.edit.fields.map(field => {
            if (field.name === e.target.name)
              return {
                ...field,
                value: e.target.value
              };
            else return field;
          })
        }
      }
    });
  };

  const submitHandler = e => {
    e.preventDefault();

    const changes = state.changes;

    const previousState = changes.prevState;

    const nextState = changes.nextState;

    let filter = {
      user: props.user.data.login,
      previousState: {},
      nextState: {}
    };

    // Проходим по полям которые могли быть изменены и сравниваем значения полей предыдущего состояния с новым,
    // после записываем данные в массив и готово
    if (nextState.name && nextState.name !== previousState.name)
      filter = {
        ...filter,
        previousState: {
          // Свойство должно именоваться именно так, иначе будут серьезные проблемы
          name: previousState.name
        },
        nextState: {
          name: nextState.name
        }
      };

    if (nextState.status && nextState.status !== previousState.status)
      filter = {
        ...filter,
        previousState: {
          ...filter.previousState,
          status: previousState.status.name
        },
        nextState: {
          ...filter.nextState,
          status: nextState.status
        }
      };

    if (nextState.fields)
      filter = {
        ...filter,
        previousState: {
          ...filter.previousState,
          fields: previousState.fields.filter(
            (change, index) => change.value !== nextState.fields[index].value
          )
        },
        nextState: {
          ...filter.nextState,
          fields: nextState.fields.filter(
            (change, index) =>
              change.value !== previousState.fields[index].value
          )
        }
      };

    return axios({
      method: "PATCH",
      url: `/api/orders/`,
      params: {
        id: state.edit._id
      },
      data: {
        ...state.edit,
        changes: [...state.edit.changes, { ...filter }]
      }
    })
      .then(res =>
        setState({
          ...state,
          response: {
            ...initialState.response,
            success: {
              status: res.status,
              message: `Ваша заявка успешно отредактирована`
            }
          },
          edit: {
            ...res.data
          },
          changes: {
            prevState: {
              ...res.data
            },
            nextState: {}
          }
        })
      )
      .catch(err =>
        setState({
          ...initialState,
          response: {
            ...initialState.response,
            error: {
              status: err.response.status,
              message: `Oops, что то пошло не так.`
            }
          }
        })
      )
      .then(() => props.update());
  };

  const changeComment = e => {
    return setState({
      ...state,
      comment: e.target.value
    });
  };

  const submitComment = e => {
    e.preventDefault();

    return axios({
      method: "PATCH",
      url: `/api/orders/`,
      params: {
        id: state.edit._id
      },
      data: {
        comments: [
          ...state.edit.comments,
          {
            user: props.user.data.login,
            content: state.comment
          }
        ]
      }
    })
      .then(res =>
        setState({
          ...state,
          response: {
            ...initialState.response,
            success: {
              status: res.status,
              message: `Ваш комментарий успешно отправлен`
            }
          },
          edit: {
            ...state.edit,
            comments: [...res.data.comments]
          },
          comment: ""
        })
      )
      .catch(err =>
        setState({
          ...state,
          response: {
            ...initialState.response,
            error: {
              status: err.response.status,
              message: `Oops, что то пошло не так.`
            }
          }
        })
      )
      .then(() => props.update());
  };

  const comments = [...state.edit.changes, ...state.edit.comments]
    .sort((a, b) => Date.parse(a.data) - Date.parse(b.data))
    .map((comment, index) => {
      if (comment.nextState) {
        if (comment.nextState.name) {
          return (
            <ListItem key={index} className={classes.listItem}>
              <ListItemText
                primary={`Пользователь ${comment.user} внес изменения в имени заявки`}
                secondary={
                  <React.Fragment>
                    <Typography
                      component="span"
                      variant="body2"
                      color="textPrimary"
                    ></Typography>
                    {` — ${comment.previousState.name} -> ${comment.nextState.name}`}
                    <small className={classes.data}>{comment.data}</small>
                  </React.Fragment>
                }
              />
            </ListItem>
          );
        }
        if (comment.nextState.status) {
          return (
            <ListItem key={index} className={classes.listItem}>
              <ListItemText
                primary={`Пользователь ${comment.user} внес изменения в поле статус`}
                secondary={
                  <React.Fragment>
                    <Typography
                      component="span"
                      variant="body2"
                      color="textPrimary"
                    ></Typography>
                    {` — ${comment.previousState.status} -> ${comment.nextState.status}`}
                    <small className={classes.data}>{comment.data}</small>
                  </React.Fragment>
                }
              />
            </ListItem>
          );
        }
        if (comment.nextState.fields) {
          return comment.nextState.fields.map((field, index) => (
            <ListItem key={index} className={classes.listItem}>
              <ListItemText
                primary={`Пользователь ${comment.user} внес изменения в поле ${field.name}:`}
                secondary={
                  <React.Fragment>
                    <Typography
                      component="span"
                      variant="body2"
                      color="textPrimary"
                    ></Typography>
                    {` — ${comment.previousState.fields[index].value} -> ${field.value}`}
                    <small className={classes.data}>{field.data}</small>
                  </React.Fragment>
                }
              />
            </ListItem>
          ));
        }
      } else {
        return (
          <ListItem key={index} className={classes.listItem}>
            <ListItemText
              primary={`Пользователь ${comment.user} оставил комментарий`}
              secondary={
                <React.Fragment>
                  <Typography
                    component="span"
                    variant="body2"
                    color="textPrimary"
                  ></Typography>
                  {` — ${comment.content}`}
                  <small className={classes.data}>{comment.data}</small>
                </React.Fragment>
              }
            />
          </ListItem>
        );
      }

      return false;
    });

  const commentsList = React.createRef();

  useEffect(() => {
    commentsList.current.scrollTop = commentsList.current.scrollHeight;
  });

  const fields = state.edit.fields.map((field, index) => (
    <TextField
      className={classes.textField}
      key={index}
      type="text"
      label={field.name}
      name={field.name}
      value={field.value}
      variant="outlined"
      onChange={changeFieldsHandler}
      fullWidth={true}
    />
  ));

  const statuses = props.statuses.map((status, index) => (
    <MenuItem id={status._id} key={index} value={status.name}>
      {status.name}
    </MenuItem>
  ));

  return (
    <Box className={classes.wrapper}>
      <form onSubmit={submitHandler}>
        <Container>
          {state.response.error.status ? (
            <Typography
              color="error"
              className={classes.error}
              variantMapping={{ h3: "h3" }}
            >
              {state.response.error.message}
            </Typography>
          ) : (
            false
          )}
          {state.response.success.status ? (
            <Typography
              color="error"
              className={classes.success}
              variantMapping={{ h3: "h3" }}
            >
              {state.response.success.message}
            </Typography>
          ) : (
            false
          )}
          <Typography className={classes.title} variant="h2" noWrap>
            Редактирование
          </Typography>
          <TextField
            label="Название заявки"
            type="text"
            name="name"
            value={state.edit.name}
            margin="normal"
            variant="outlined"
            onChange={changeNameHandler}
            fullWidth={true}
            className={classes.textField}
          />
          <Divider className={classes.divider} />
          <Typography className={classes.title} variant="h3" noWrap>
            Статус заявки
          </Typography>
          <Select
            id="demo-simple-select-outlined"
            value={state.edit.status.name}
            variant="outlined"
            onChange={changeStatusHandler}
            fullWidth={true}
            inputProps={{ name: "status" }}
          >
            {statuses}
          </Select>
          <Divider className={classes.divider} />
          <Typography className={classes.title} variant="h3" noWrap>
            Произвольные поля
          </Typography>
          {fields}
          <Divider className={classes.divider} />
          {state.edit.name.length < 1 ? (
            <Button variant="contained" disabled>
              Редактировать заявку
            </Button>
          ) : (
            <Button variant="contained" color="primary" type="submit">
              Редактировать заявку
            </Button>
          )}
        </Container>
      </form>
      <Container>
        <Typography className={classes.title} variant="h2" noWrap>
          Комментарии
        </Typography>
        <List className={classes.list} ref={commentsList}>
          {comments}
        </List>
        <form className={classes.comment} onSubmit={submitComment}>
          <TextField
            id="name"
            type="text"
            label="Комментарий"
            name="comment"
            margin="dense"
            value={state.comment}
            onChange={changeComment}
            fullWidth
          />
          <Divider className={classes.divider} />
          {state.comment.length < 1 ? (
            <Button variant="contained" disabled>
              Добавить
            </Button>
          ) : (
            <Button variant="contained" color="primary" type="submit">
              Добавить
            </Button>
          )}
        </form>
      </Container>
    </Box>
  );
};

const mapStateToProps = state => {
  return {
    user: state.authorization.payload,
    statuses: state.statuses.payload
  };
};

const mapDispatchToProps = dispatch => {
  return {
    update: () => dispatch(fetchUpdateOrderList())
  };
};

EditOrder.propTypes = {
  user: PropTypes.object.isRequired,
  update: PropTypes.func.isRequired,
  statuses: PropTypes.array.isRequired,
  editOrder: PropTypes.object.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(EditOrder);
