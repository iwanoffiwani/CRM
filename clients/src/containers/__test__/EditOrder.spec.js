import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import fetchMock from 'fetch-mock'
import assert from 'assert'

import { EditOrder } from '../EditOrder'
import TextField from '@material-ui/core/TextField'
import Select from '@material-ui/core/Select'
import MenuItem from '@material-ui/core/MenuItem'

describe('Tests EditOrder component', () => {
  const props = {
    editOrder: {
      "_id": "5e28371aa7b3b8210c9d38de",
      "name": "Новая заявка-35",
      "user": "5da47a1353265a14c80ef289",
      "fields": [
        {
          "value": "+7 666 666 66 66",
          "_id": "5dcac8a093dd3b05f8cb0adc",
          "name": "Мобильный телефон"
        },
        {
          "value": "79 98 666 555",
          "_id": "5df9f19e1607be20381256e8",
          "name": "Номер паспорта"
        },
        {
          "value": "516 69 69 ",
          "_id": "5dfa29e98ef6f51a186c395e",
          "name": "Домашний тел"
        },
        {
          "value": "79 98 654 654",
          "_id": "5e00b1bfa63229263405fd21",
          "name": "Паспорт"
        }
      ],
      "status": {
        "_id": "5e28379ffa502c1edc0e768c",
        "name": "Статус-4"
      },
      "comments": [],
      "changes": [
        {
          "data": "2020-01-22T11:53:27.209Z",
          "_id": "5e2837b7fa502c1edc0e768e",
          "user": "test@test",
          "previousState": {
            "status": "None"
          },
          "nextState": {
            "status": "Статус-2"
          }
        }
      ]
    }
  }

  const mapStateToProps = {
    statuses: [
      {
        "_id": "5e28378cfa502c1edc0e7689",
        "name": "Статус-1"
      },
      {
        "_id": "5e283799fa502c1edc0e768b",
        "name": "Статус-2"
      },
      {
        "_id": "5e28379ffa502c1edc0e768c",
        "name": "Статус-4"
      },
      {
        "_id": "5e2837a5fa502c1edc0e768d",
        "name": "Статус-5"
      }
    ],
    user: {
      data: {
        login: "test@test"
      }
    }
  }

  const mapDispatchToProps = {
    update: jest.fn()
  }

  it('+++ renderer component', () => {
    const wrapper = shallow(<EditOrder {...props}{...mapStateToProps}{...mapDispatchToProps} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('+++ Checking if the order name is in the field', () => {
    const wrapper = shallow(<EditOrder {...props}{...mapStateToProps}{...mapDispatchToProps} />)

    const {
      editOrder: {
        name
      }
    } = props

    const orderLabelWrapper = wrapper.find(TextField).at(0)
    expect(orderLabelWrapper.props().value).toEqual(name)
  })

  it('+++ getting the correct status value', () => {
    const wrapper = shallow(<EditOrder {...props}{...mapStateToProps}{...mapDispatchToProps} />)

    const {
      editOrder: {
        status: {
          name
        }
      }
    } = props

    const statusOrderWrapper = wrapper.find(Select).at(0)
    expect(statusOrderWrapper.props().value).toEqual(name)
  })

  it('+++ check for statuses in the order', () => {
    const wrapper = shallow(<EditOrder {...props}{...mapStateToProps}{...mapDispatchToProps} />)

    const {
      statuses
    } = mapStateToProps

    const statusesOrderWrapper = wrapper.find(MenuItem)
    expect(statusesOrderWrapper.at(0).props().value).toEqual(statuses[0].name)
  })

  it('+++ simulation of order status change', async () => {
    const wrapper = shallow(<EditOrder {...props}{...mapStateToProps}{...mapDispatchToProps} />)

    const {
      editOrder: {
        status: {
          name, _id
        }
      }
    } = props

    const {
      statuses
    } = mapStateToProps

    const selectOrderWrapper = wrapper.find(Select).at(0)
    expect(selectOrderWrapper.props().value).toEqual(name)

    const nextStatus = name === statuses[0].name ? statuses[1].name : statuses[0].name  
    const nextStatusID = _id === statuses[0]._id ? statuses[1]._id : statuses[0]._id  

    selectOrderWrapper.simulate('change', { 
      _targetInst: {
        stateNode: {
          id: nextStatusID
        }
      },
      target: { 
        value: nextStatus 
      } 
    })

    expect(wrapper.find(Select).at(0).props().value).toEqual(nextStatus)
  })

  it('+++ check for fields in the order', () => {
    const wrapper = shallow(<EditOrder {...props}{...mapStateToProps}{...mapDispatchToProps} />)

    const {
      editOrder: {
        fields
      }
    } = props

    // Если первое поле получило верное значение, значит цикл отработал правильно.
    // Поля заявки должны часто изменяются, по этому проверяем по первому объекту в массиве.
    const orderFieldsWrapper = wrapper.find(TextField).at(1)
    expect(orderFieldsWrapper.props().value).toEqual(fields[0].value)
  })

  it('+++ simulation of changing field data', () => {
    const wrapper = shallow(<EditOrder {...props}{...mapStateToProps}{...mapDispatchToProps} />)

    const {
      editOrder: {
        fields
      }
    } = props

    const orderFieldsWrapper = wrapper.find(TextField).at(1)
    expect(orderFieldsWrapper.props().value).toEqual(fields[0].value)

    const nextValue = 'Some new value'
    
    orderFieldsWrapper.simulate('change', { 
      target: { 
        value: nextValue,
        name: orderFieldsWrapper.props().name,
      } 
    })

    expect(wrapper.find(TextField).at(1).props().value).toEqual(nextValue)
  })

  it('+++ fetch an application with changed fields and status (without displaying changes)', async () => {
    const url = new URL('http://localhost:5000/api/orders')
    const params = { id: props.editOrder._id }

    url.search = new URLSearchParams(params).toString()

    fetchMock.mock(url, 200)

    const res = await fetch(url, {
      method: 'PATCH',
      body: JSON.stringify({
        "_id": "5e28371aa7b3b8210c9d38de",
        "name": "Новая заявка-35",
        "user": "5da47a1353265a14c80ef289",
        "fields": [
          {
            "value": "+7 999 999 99 99",
            "_id": "5dcac8a093dd3b05f8cb0adc",
            "name": "Мобильный телефон"
          },
          {
            "value": "79 98 666 665",
            "_id": "5df9f19e1607be20381256e8",
            "name": "Номер паспорта"
          },
          {
            "value": "516 69 69 ",
            "_id": "5dfa29e98ef6f51a186c395e",
            "name": "Домашний тел"
          },
          {
            "value": "79 98 654 654",
            "_id": "5e00b1bfa63229263405fd21",
            "name": "Паспорт"
          }
        ],
        "status": {
          "_id": "5e28378cfa502c1edc0e7689",
          "name": "Статус-1"
        },
        "comments": [],
        "changes": [
          {
            "data": "2020-01-22T11:53:27.209Z",
            "_id": "5e2837b7fa502c1edc0e768e",
            "user": "test@test",
            "previousState": {
              "status": "None"
            },
            "nextState": {
              "status": "Статус-2"
            }
          }
        ]
      })
    })

    assert(res)
    fetchMock.restore()
  })

  it('+++ simulation of changing the comment input field', () => {
    const wrapper = shallow(<EditOrder {...props}{...mapStateToProps}{...mapDispatchToProps} />)
    
    const textFieldWrapper = wrapper.find(TextField).last()
    expect(textFieldWrapper.props().value).toEqual('')

    const comment = 'Some comments'

    textFieldWrapper.simulate('change', {
      target: {
        value: comment
      }
    })

    expect(wrapper.find(TextField).last().props().value).toEqual(comment)
  })

  it('+++ fetch a new comment', async () => {
    const url = new URL('http://localhost:5000/api/orders')
    const params = { id: props.editOrder._id }

    url.search = new URLSearchParams(params).toString()

    fetchMock.mock(url, 200)
      
    const res = await fetch(url, {
      method: 'PATCH',
      body: JSON.stringify({
        "comments": [{
          user: 'test@test',
          content: 'Some new comment'
        }]
      })
    })

    assert(res)
    fetchMock.restore()
  })
})