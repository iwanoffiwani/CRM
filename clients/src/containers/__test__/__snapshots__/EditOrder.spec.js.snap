// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests EditOrder component +++ renderer component 1`] = `
<Styled(MuiBox)
  className="makeStyles-wrapper-1"
>
  <form
    onSubmit={[Function]}
  >
    <WithStyles(ForwardRef(Container))>
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-title-6"
        noWrap={true}
        variant="h2"
      >
        Редактирование
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(TextField))
        className="makeStyles-textField-5"
        fullWidth={true}
        label="Название заявки"
        margin="normal"
        name="name"
        onChange={[Function]}
        type="text"
        value="Новая заявка-35"
        variant="outlined"
      />
      <WithStyles(ForwardRef(Divider))
        className="makeStyles-divider-7"
      />
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-title-6"
        noWrap={true}
        variant="h3"
      >
        Статус заявки
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Select))
        fullWidth={true}
        id="demo-simple-select-outlined"
        inputProps={
          Object {
            "name": "status",
          }
        }
        onChange={[Function]}
        value="Статус-4"
        variant="outlined"
      >
        <WithStyles(ForwardRef(MenuItem))
          id="5e28378cfa502c1edc0e7689"
          key="0"
          value="Статус-1"
        >
          Статус-1
        </WithStyles(ForwardRef(MenuItem))>
        <WithStyles(ForwardRef(MenuItem))
          id="5e283799fa502c1edc0e768b"
          key="1"
          value="Статус-2"
        >
          Статус-2
        </WithStyles(ForwardRef(MenuItem))>
        <WithStyles(ForwardRef(MenuItem))
          id="5e28379ffa502c1edc0e768c"
          key="2"
          value="Статус-4"
        >
          Статус-4
        </WithStyles(ForwardRef(MenuItem))>
        <WithStyles(ForwardRef(MenuItem))
          id="5e2837a5fa502c1edc0e768d"
          key="3"
          value="Статус-5"
        >
          Статус-5
        </WithStyles(ForwardRef(MenuItem))>
      </WithStyles(ForwardRef(Select))>
      <WithStyles(ForwardRef(Divider))
        className="makeStyles-divider-7"
      />
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-title-6"
        noWrap={true}
        variant="h3"
      >
        Произвольные поля
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(TextField))
        className="makeStyles-textField-5"
        fullWidth={true}
        key="0"
        label="Мобильный телефон"
        name="Мобильный телефон"
        onChange={[Function]}
        type="text"
        value="+7 666 666 66 66"
        variant="outlined"
      />
      <WithStyles(ForwardRef(TextField))
        className="makeStyles-textField-5"
        fullWidth={true}
        key="1"
        label="Номер паспорта"
        name="Номер паспорта"
        onChange={[Function]}
        type="text"
        value="79 98 666 555"
        variant="outlined"
      />
      <WithStyles(ForwardRef(TextField))
        className="makeStyles-textField-5"
        fullWidth={true}
        key="2"
        label="Домашний тел"
        name="Домашний тел"
        onChange={[Function]}
        type="text"
        value="516 69 69 "
        variant="outlined"
      />
      <WithStyles(ForwardRef(TextField))
        className="makeStyles-textField-5"
        fullWidth={true}
        key="3"
        label="Паспорт"
        name="Паспорт"
        onChange={[Function]}
        type="text"
        value="79 98 654 654"
        variant="outlined"
      />
      <WithStyles(ForwardRef(Divider))
        className="makeStyles-divider-7"
      />
      <WithStyles(ForwardRef(Button))
        color="primary"
        type="submit"
        variant="contained"
      >
        Редактировать заявку
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Container))>
  </form>
  <WithStyles(ForwardRef(Container))>
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-title-6"
      noWrap={true}
      variant="h2"
    >
      Комментарии
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(List))
      className="makeStyles-list-10"
    >
      <WithStyles(ForwardRef(ListItem))
        className="makeStyles-listItem-11"
        key="0"
      >
        <WithStyles(ForwardRef(ListItemText))
          primary="Пользователь test@test внес изменения в поле статус"
          secondary={
            <React.Fragment>
              <ForwardRef(WithStyles)
                color="textPrimary"
                component="span"
                variant="body2"
              />
               — None -&gt; Статус-2
              <small
                className="makeStyles-data-12"
              >
                2020-01-22T11:53:27.209Z
              </small>
            </React.Fragment>
          }
        />
      </WithStyles(ForwardRef(ListItem))>
    </WithStyles(ForwardRef(List))>
    <form
      className="makeStyles-comment-9"
      onSubmit={[Function]}
    >
      <WithStyles(ForwardRef(TextField))
        fullWidth={true}
        id="name"
        label="Комментарий"
        margin="dense"
        name="comment"
        onChange={[Function]}
        type="text"
        value=""
      />
      <WithStyles(ForwardRef(Divider))
        className="makeStyles-divider-7"
      />
      <WithStyles(ForwardRef(Button))
        disabled={true}
        variant="contained"
      >
        Добавить
      </WithStyles(ForwardRef(Button))>
    </form>
  </WithStyles(ForwardRef(Container))>
</Styled(MuiBox)>
`;
